update test framework and tests to support 4x version of Pester

JamesWTruher opened the pull request:
Use the latest version of Pester from the Gallery to invoke our tests. A number of test changes needed to be made because some assertions were dropped in the 4x version of Pester.
I have left the Restore-PSPester function in the case that someone would like to have the older version. This can certainly be removed if desired.

JamesWTruher committed:
Test fixes and changes needed to support Pester 4.0.8

JamesWTruher committed:
Replace 'Should Contain' with new 'Should FileContentMatch' assertion

Explicitly check for string creation with write-output

JamesWTruher committed:
Use the current version of pester and install it in modulesDir

JamesWTruher committed:
Simplify logic for relative path test.

Multiple '..' is not needed for a relative path, a single one will do. Also, on multi-drive systems using split-path -noqualifier will probably do the wrong thing with regard to constructing a correct path.
Remove extraneous Should Not Throw test, if this throws, the test will fail, we don't need to explicitly assert the not throw

markekraus commented:
Maintainers: please make sure this passes `[feature]` tests before merging. 

@JamesWTruher can you please push a commit with `[feature]` in the commit message? Thanks!

JamesWTruher committed:
In some environments it is possible that computer name is 'localhost', so that should be allowed

stknohg commented:
Why don’t we adopt Pester 4.0 new assertion syntax?

e.g.

```powershell
# current syntax
$var1 | Should Not Be 10

# Pester 4.0 new assertion syntax
$var1 | Should -Not -Be 10
```

markekraus commented:
@stknohg My preference would be to make the change to the new assertion syntax in separate PR's, one per test file, after we have successfully moved the tests to work with 4.0. It's one thing to make the tests work with 4.0, it is an entirely other thing to switch to the new syntax. switching all tests to the new syntax in this PR would make it hard to see what changes were necessary just to get it working with 4.0.

stknohg commented:
@markekraus I think it's good. 👍

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1:
```diff
@@ -108,73 +108,73 @@ Describe "Remove-Variable" -Tags "CI" {
+	    { Remove-Variable -Name var1 -Scope local -ErrorAction Stop } | Should Throw
 
-	    $var1 | Should Be # Nothing
+	    $var1 | Should be context
```
be -> Be

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1:
```diff
@@ -67,8 +67,8 @@ set-psbreakpoint -command foo
         $brk = sbp -line 11,12,13 -column 1 -script $scriptFileName,$scriptFileName
-        $brk.Line | Should Be 11,12,13
-        $brk.Column | Should Be 1
+        $brk.Line | Should BeIn 11,12,13
```
Above I see another pattern like `Should BeIn @(11,12,13)`. 
What pattern we'll use in tests?

JamesWTruher replied:
I don't think it matters - both are valid expressions

SteveL-MSFT replied:
We should probably just standardize on one of them for consistency/readability.  I prefer sans @(), myself.

iSazonov replied:
I agree that we will request `@()` template for `Should BeIn` in future test reviews.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:
```diff
@@ -42,7 +42,7 @@
     It "should only load the specified version" {
         Import-Module TestModule -RequiredVersion 1.1
-        (Get-Module TestModule).Version | Should Be "1.1"
+        (Get-Module TestModule).Version | Should BeIn "1.1"
```
Isn't `BeIn` only used for validating if the left side is contained in the collection on the right side?  Don't understand why this isn't just `Be`?

JamesWTruher replied:
as it turns out, there are multiple loaded modules which are named TestModule, thus a comparison with "Be" fails because it's a collection. I didn't want to rearchitect the tests.

SteveL-MSFT replied:
Got it. Closed.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:
```diff
@@ -192,7 +192,7 @@ Describe "Import-Module should be case insensitive" -Tags 'CI' {
         $m = Get-Module TESTmodule
         $m | Should BeOfType "System.Management.Automation.PSModuleInfo"
-        $m.Name | Should Be "TESTMODULE"
+        $m.Name | Should BeIn "TESTMODULE"
```
Isn't BeIn only used for validating if the left side is contained in the collection on the right side? Don't understand why this isn't just Be?

JamesWTruher replied:
same as above

SteveL-MSFT replied:
Closed

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1:
```diff
@@ -145,7 +145,8 @@ Describe "Get-Variable" -Tags "CI" {
 
-	    Get-Variable -Name globalVar -Scope local -ErrorAction SilentlyContinue | Should Throw
+	    Get-Variable -Name globalVar -Scope local -ErrorAction SilentlyContinue -ErrorVariable removeGlobalAsLocal
+        $removeGlobalAsLocal.FullyQualifiedErrorId | Should Be "VariableNotFound,Microsoft.PowerShell.Commands.GetVariableCommand"
```
Is this line indented correctly?

JamesWTruher replied:
there is a sad combination of tabs and spaces in this file, I'll fix up this line and save the whitespace fixing for a later PR

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1:
```diff
@@ -5,7 +5,7 @@ Describe "Remove-Variable" -Tags "CI" {
 	New-Variable -Name var1 -Value 4
 
-	Remove-Variable $var1 -ErrorAction SilentlyContinue | Should Throw
+	{ Remove-Variable $var1 -ErrorAction Stop } | Should Throw
```
Wouldn't it be better to change this to:

```powershell
Remove-Variable $var1 -ErrorAction SilentlyContinue -ErrorVariable err
$err.FullyQualifiedErrorId | Should Be "VariableNotFOund,Microsoft.PowerShell.Commands.RemoveVariableCommand"
```

(applies to all the `Should throw` checks)

JamesWTruher replied:
I'll update the changes that I made to get it to work in this way, but I will defer changing all of these to a follow-on PR. My plan is to not update all tests with this PR

SteveL-MSFT replied:
Ok.  Closed.

bergmeister commented:
Can we please update one of the readme files with basic instructions and a link how to upgrade pester v3 to v4 (since this is not straightforward in most cases)?

JamesWTruher committed:
[feature] Add link for migrating tests from Pester v3 to v4

Fix up capitalization and white space issues
Change one test to check FullyQualifiedErrorId rather than just `Should Throw`

JamesWTruher committed:
[feature] update invoke-item test to handle the case where multiple notepad processes are running

iSazonov commented on docs/testing-guidelines/WritingPesterTests.md:
```diff
@@ -1,7 +1,11 @@
+Note that this document does not replace the documents found in the [Pester](https://github.com/pester/pester "Pester") project.
+This is just some quick tips and suggestions for creating Pester tests for this project.
+The Pester community is vibrant and active, if you have questions about Pester or creating tests, the [Pester Wiki](https://github.com/pester/pester/wiki) has a lot of great information.
+As of January 2018, PowerShell Core is using Pester version 4x which has some changes from earlier versions.
```
Speller test failed on "4x"

daxian-dbw replied:
The corresponding error in Travis CI:
```
The command "pwsh -File tools/travis.ps1" exited with 0.
2.37s$ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then mdspell '**/*.md' '!**/Pester/**/*.md' --ignore-numbers --ignore-acronyms --report; fi
    docs/testing-guidelines/WritingPesterTests.md
        5 |  Core is using Pester version 4x which has some changes from e 
>> 1 spelling error found in 67 files
```

iSazonov commented:
CI Apveyor temporary failed on Web tests. #6069 address the problem.

JamesWTruher committed:
Fix spelling issue with Pester 4x, calling it Pester 4 should be sufficient

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Item.Tests.ps1:
```diff
@@ -67,7 +67,8 @@ Describe "Invoke-Item basic tests" -Tags "Feature" {
                 $notepadProcess = Get-Process -Name $notepadProcessName
-                $notepadProcess.Name | Should Be $notepadProcessName
+                # we need BeIn because multiple notepad processes could be running
+                $notepadProcess.Name | Should BeIn $notepadProcessName
```
In line 67 we kill all notepad processes.

JamesWTruher replied:
in CI this is probably ok - on a dev box, not so much - the fix for this should be part of the generalized test review

iSazonov reviewed (approved):
LGTM with one comment.
